gui/biomes.lua

```diff
@@ -0,0 +1,388 @@
+    end
+end
+
+local function add_field_text(lines, biome, field_name)
+    lines[#lines+1] = ("%s: %s"):format(field_name, biome[field_name])
+    lines[#lines+1] = NEWLINE
+end
+
+local function get_tooltip_text(option)
+    if not option then
+        return ""
+    end
+
+    local text = {}
+    text[#text+1] = ("type: %s"):format(df.biome_type[option.biomeTypeId])
+    text[#text+1] = NEWLINE
+
+    local biome = option.biome
+
+    add_field_text(text, biome, "savagery")
+    add_field_text(text, biome, "evilness")
+    table.insert(text, NEWLINE)
+
+    add_field_text(text, biome, "elevation")
+    add_field_text(text, biome, "rainfall")
+    add_field_text(text, biome, "drainage")
+    add_field_text(text, biome, "vegetation")
+    add_field_text(text, biome, "temperature")
+    add_field_text(text, biome, "volcanism")
+    table.insert(text, NEWLINE)
+
+    local flags = biome.flags
+    if flags.is_lake then
+        text[#text+1] = "lake"
+        text[#text+1] = NEWLINE
+    end
+    if flags.is_brook then
+        text[#text+1] = "brook"
+        text[#text+1] = NEWLINE
+    end
+
+    return text
+end
+
+function BiomeVisualizerLegend:onMouseHoverEntry(idx, option)
+    self:ShowTooltip(option or self.SelectedOption)
+end
+
+function BiomeVisualizerLegend:ShowTooltip(option)
+    local text = get_tooltip_text(option)
+
+    local tooltip_panel = self.tooltip_panel
+    local lbl = tooltip_panel.subviews.label
+
+    lbl:setText(text)
+end
+
+--------------------------------------------------------------------------------
+
+if RELOAD then BiomeVisualizer = nil end
+BiomeVisualizer = defclass(BiomeVisualizer, gui.ZScreen)
+BiomeVisualizer.ATTRS{
+    focus_path='BiomeVisualizer',
+    pass_movement_keys=true,
+}
+
+function BiomeVisualizer:init()
+    local legend = BiomeVisualizerLegend{view_id = 'legend'}
+    self:addviews{legend}
+end
+
+function BiomeVisualizer:onRenderFrame(dc, rect)
+    BiomeVisualizer.super.onRenderFrame(self, dc, rect)
+
+    if not dfhack.screen.inGraphicsMode() and not gui.blink_visible(500) then
+        return
+    end
+
+    local z = df.global.window_z
+    if not biomesMap[z] then
+        gatherBiomeInfo(z)
+        self.subviews.legend:UpdateChoices()
+    end
+
+    local function get_overlay_pen(pos)
+        local self = self
+        local safe_index = safe_index
+        local biomes = biomesMap
+
+        local N = safe_index(biomes, pos.z, pos.y, pos.x)
+        if not N then return end
+
+        local idxSelected = self.subviews.legend.SelectedIndex
+        local idxTile = (N == idxSelected)
+                    and TILE_HIGHLIGHTED
+                    or TILE_STARTING_SYMBOL + (N-1)
+        local color = (N == idxSelected)
+                    and COLOR_CYAN
+                    or COLOR_GREY
+        local ch = string.char(string.byte('a') + (N-1))
+        return color, ch, idxTile
+    end
+
+    guidm.renderMapOverlay(get_overlay_pen, nil) -- nil for bounds means entire viewport
```

Comment:
I'm amazed that is as performant as it is. When I was developing `gui/quickfort`, rendering to the entire viewport was palpably slow, but this seems to run without any delay.

edit: `gui/quickfort` is still kind of slow. I guess it's the algorithm there that needs improvement.